src/proposed/0140-budget-limit-threshold.md

```diff
@@ -0,0 +1,469 @@
+
+The second stage of budget limits is motivated by the current implicit
+requirement that clients must completely trust passive servers that they call.
+Under the current model, a server can consume as much budget as it desires on a
+donated SC and is not forced to return it. A malfunctioning or malicious passive
+server can therefore run indefinitely while also possessing a higher priority
+than its clients, significantly impeding their ability to make progress.
+
+The impact of this extends to any thread with a priority between that of the
+passive server and its clients, even threads that are not themselves clients of
+the passive server. Enforcing stronger bounds on passive server SC consumption
+would reduce the level of trust threads in a system need to have in one another.
+
+## Guide-level explanation
+
+### Thresholds Guide
+
+This change involves adding a new attribute to endpoint objects, known as a
+threshold. By default, this threshold value would be set to zero and will have
+no effect. Therefore, existing code will not need to be updated to accommodate
+thresholds.
+
+If the threshold is set to a non-zero value, sending invocations performed on
+the endpoint will only be permitted if they donate a scheduling context with
+available budget greater than or equal to the threshold. Available budget is the
+amount of budget ready for use in the SC (the sum of refills with a release time
+greater than the current time). By setting the threshold to the WCET of the
+passive server associated with the endpoint, this guarantees that the server
+will always receive an SC with sufficient budget to complete its execution,
+preventing budget expiry from occurring. This will also ensure that timeout
+exceptions will only occur as a result of a true error, such as a misconfigured
+threshold value or a fault in the server code.
+
+If a client has insufficient available budget in its SC, its refills will be
+deferred and merged until its head refill has sufficient budget at some point in
+the future. Once the thread resumes, the IPC invocation will transparently
+continue. If a client's maximum budget is less than the threshold value, an
+error will instead be returned to the client.
+
+The time required to check an SC's available budget and to defer and merge are
+both bounded by the maximum number of refills in an SC. When designing a system
+using an endpoint with a threshold, this will be an additional factor to
+consider regarding the optimal choice for the maximum number of SC refills.
+
+The threshold value of an endpoint can be set via an object invocation,
+performed indirectly through an invocation on the CNode, similar to the existing
+Cancel Badged Sends invocation. However, we restrict the right to set the
+endpoint threshold to only the original unbadged capability.
+
+Changes to the threshold value are propagated weakly, meaning that clients only
+interact with the threshold value at the moment where they invoke the endpoint.
```

Comment:
question: There's also an implication here that all of the server's operations are bounded by the same amount. How practical is that?